src/celengine/render.cpp

```diff
@@ -7918,3 +7970,93 @@ bool Renderer::captureFrame(int x, int y, int w, int h, Renderer::PixelFormat fo
 
     return glGetError() == GL_NO_ERROR;
 }
+
+static void drawRectangle(const Renderer &renderer, const Rectangle &r)
+{
+    uint32_t p = r.tex == nullptr ? 0 : ShaderProperties::HasTexture;
+    switch (r.nColors)
+    {
+    case 0:
+        break;
+    case 1:
+        p |= ShaderProperties::UniformColor;
+        break;
+    case 4:
+        p |= ShaderProperties::PerVertexColor;
+        break;
+    default:
+        fmt::fprintf(cerr, "Incorrect number of colors: %i\n", r.nColors);
+    }
+
+    auto prog = renderer.getShaderManager().getShader(ShaderProperties(p));
+    if (prog == nullptr)
+        return;
+
+    constexpr array<short, 8> texels = {0, 1,  1, 1,  1, 0,  0, 0};
+    array<float, 8> vertices = { r.x, r.y,  r.x+r.w, r.y, r.x+r.w, r.y+r.h, r.x, r.y+r.h };
+
+    auto s = static_cast<GLsizeiptr>(memsize(vertices) + memsize(texels) + 4*4*sizeof(GLfloat));
+    static celgl::VertexObject vo{ GL_ARRAY_BUFFER, s, GL_DYNAMIC_DRAW };
+    vo.bindWritable();
+
+    if (!vo.initialized())
+    {
+        vo.allocate();
+        vo.setBufferData(texels.data(), memsize(vertices), memsize(texels));
+        vo.setVertices(2, GL_FLOAT);
+        vo.setTextureCoords(2, GL_SHORT, false, 0, memsize(vertices));
+        vo.setColors(4, GL_FLOAT, false, 0, memsize(vertices) + memsize(texels));
+    }
+
+    vo.setBufferData(vertices.data(), 0, memsize(vertices));
+    if (r.nColors == 4)
+    {
+        array<Vector4f, 4> ct;
+        for (size_t i = 0; i < 4; i++)
+            ct[i] = r.colors[i].toVector4();
+        vo.setBufferData(ct.data(), memsize(vertices) + memsize(texels), 4*4*sizeof(GLfloat));
+    }
+
+    prog->use();
+    if (r.tex != nullptr)
+    {
+        glEnable(GL_TEXTURE_2D);
+        r.tex->bind();
+        prog->samplerParam("tex") = 0;
+    }
+
+    if (r.nColors == 1)
+        prog->vec4Param("color") = r.colors[0].toVector4();
+
+    if (r.type != Rectangle::Type::BorderOnly)
+    {
+        vo.draw(GL_TRIANGLE_FAN, 4);
+    }
+    else
+    {
+        if (r.lw != 1.0f)
+            glLineWidth(r.lw);
+        vo.draw(GL_LINE_LOOP, 4);
+        if (r.lw != 1.0f)
+            glLineWidth(1.0f);
+    }
+
+    glUseProgram(0);
+    vo.unbind();
+}
+
+void Renderer::drawRectangle(const Rectangle &r) const
+{
+    ::drawRectangle(*this, r);
```

Comment:
> Why separate static function just as a method content?

Because it has some static data and I don't need to put them to `Renderer`, it's already has a lot of crap inside and requires refactoring. So until we designed a proper new `Renderer` I'd prefer not to introduce a lot of changes to it.